README.md
# @flowforge/nr-persistent-context

A Node-RED Context Plugin for the FlowForge platform.

This plugin provides persistent context for a Node-RED instance
on the FlowForge platform.

### Configuration

```js
contextStorage: {
    file: {
        module: require("@flowforge/nr-persistent-context"),
        config: {
            projectID: process.env['FORGE_PROJECT_ID'],
            baseURL: process.env['FORGE_STORAGE_URL'],
            token: process.env['FORGE_STORAGE_TOKEN'],
            requestTimeout: 3000,
            pageSize: 20,
            flushInterval: 30,
            cache: true
        }
    }
}
```

 - `projectID` - is the UUID of the project (provided by FlowForge)
 - `baseURL` - the root URL for the FlowForge Storage API (provided by FlowForge)
 - `token` - authentication token (provided by FlowForge)
 - `requestTimeout` - (optional) The number of milliseconds to wait before timing out a request (Type:`number`, Default:`3000`)
 - `pageSize` - (optional) The number of context items/rows to fetch per page (Type:`number`, Default:`20`)
 - `flushInterval` - (optional) The number of seconds to wait before flushing pending writes (Type:`number`, Default:`30`)
 - `cache` - (optional) Whether to cache context items in memory (required for synchronous get/set) (Type:`boolean`, Default:`true`)


CHANGELOG.md
### 0.0.5 Release

 - Introduce build and publish pipeline (#25) @ppawlowski
 - Disable scheduled package build (#32) @ppawlowski
 - Add nr-launcher package dispatcher (#31) @ppawlowski
 - Update to NR 3.1
 - Prep for NR 3.1 (#30) @hardillb
 - apply all patch updates
 - Bump fast-xml-parser and @aws-sdk/client-s3 (#29) @app/dependabot
 - Bump word-wrap from 1.2.3 to 1.2.5 (#28) @app/dependabot
 - FIX: Publish package on schedule (#27) @ppawlowski
 - FIX: Allow publish only when changes are pushed to `main` branch (#26) @ppawlowski
 - Chore: Set root flag in eslint (#24) @Pezmc
 - Add package-lock.json (#23) @Pezmc

#### 0.0.4: Release

 - fix cache write crash (missing _writeCache func) (#20) @Steve-Mcl
 - Docs synchronous update (#21) @Steve-Mcl
 - Add cache layer and calls to new API to support synchronous context (#18) @Steve-Mcl

#### 0.0.3: Release

 - Relicense to the FF License (#13) @hardillb
 - Add limitations to readme (#12) @Steve-Mcl
 - improve errors for node-red, incl 413 payload size (#9) @Steve-Mcl
 - Add tests (#8) @Steve-Mcl
 - Add safe stringify dependency (#7) @Steve-Mcl

test/unit/setup.js
const http = require('http')
const YAML = require('yaml')

// setup mock app with authentication endpoint
// that provides auth token validation for testing
function authServer (config = {}) {
    const host = config.host || '127.0.0.1'
    const port = config.port || 3002
    const authConfig = config.authConfig || [
        { token: 'test-token-1', projectId: 'test-project-1' }
    ]
    const requestListener = function (req, res) {
        try {
            let authToken
            const urlParts = req.url.split('/')
            const projectId = urlParts.pop()
            const route = urlParts.join('/')
            switch (route) {
            case '/account/check/project':
                authToken = authConfig.find(auth => auth.projectId === projectId)
                if (req.headers.authorization === ('Bearer ' + authToken.token)) {
                    res.writeHead(200)
                    res.end('{}')
                    return
                }
                throw new Error('Unknown request')
            default:
                res.writeHead(404)
                res.end(JSON.stringify({ error: 'Resource not found' }))
            }
        } catch (error) {
            res.writeHead(401)
            res.end(JSON.stringify({ error: 'unauthorised' }))
        }
    }

    const authServer = http.createServer(requestListener)
    authServer.listen(port, host, () => {
        // listening for requests
    })
    return authServer
}

/**
 * Setup a test file server app
 * @param {Object} config test settings
 * @param {string} [config.home] (optional) path to flowforge home directory
 * @param {string} config.projectId project id
 * @param {string} config.teamId team id
 * @param {string} config.host host
 * @param {number} config.port port
 * @param {string} config.base_url base url
 * @param {Object} config.context context settings
 * @param {'memory'|'sequelize'|'redis'} config.context.type context type
 * @param {number} config.context.quota context quota (max size in MB)
 * @param {Object} config.context.options context options (depends on context type)
 * @param {'sqlite'|'postgres'} config.context.options.type context options type
 * @returns @flowforge/file-server
 */
async function setupFileServerApp (config = {}) {
    const FF_FS_TEST_CONFIG = {
        FLOWFORGE_HOME: config.home || process.cwd(),
        FLOWFORGE_PROJECT_ID: config.projectId,
        FLOWFORGE_TEAM_ID: config.teamId,
        host: config.host,
        port: config.port,
        base_url: config.base_url,
        driver: {
            type: 'memory',
            options: {
                root: 'var/root'
            }
        },
        context: {
            type: config.context?.type,
            quota: config.context?.quota || 1000,
            options: {
                type: 'sqlite',
                storage: 'ff-context.db'
            }
        }
    }
    if (config.context?.options?.type === 'postgres') {
        FF_FS_TEST_CONFIG.context.options.storage = { ...config.context?.options }
    }
    process.env.FF_FS_TEST_CONFIG = YAML.stringify(FF_FS_TEST_CONFIG)
    const app = await require('@flowforge/file-server')
    return app
}

module.exports = {
    authServer,
    setupFileServerApp
}


lib/index.js
const got = require('got').default
const safeJSONStringify = require('json-stringify-safe')
const CONFIG_ERROR_MSG = 'Persistent context plugin cannot be used outside of FlowForge EE environment'

function stringify (value) {
    let hasCircular
    const result = safeJSONStringify(value, null, null, function (k, v) { hasCircular = true })
    return { json: result, circular: hasCircular }
}

const reviver = (keys, data) => {
    const result = keys.map(key => {
        const el = data.find(e => e.key === key)
        return el?.value
    })
    return result
}

function eeRejectOrCallback (reject, callback) {
    if (callback) {
        callback(new Error(CONFIG_ERROR_MSG))
    } else if (reject) {
        return Promise.reject(new Error(CONFIG_ERROR_MSG))
    }
}

function normaliseError (err) {
    let niceError = new Error('Unexpected error.')
    let statusCode = null
    let childErr = {}
    niceError.code ||= 'unexpected_error'
    if (typeof err === 'string') {
        niceError = new Error(err)
    } else if (err?._normalised) {
        return err // already normalised
    }
    if (err?.response) {
        statusCode = err.response.statusCode
        if (err.response.body) {
            try {
                if (err.response.body && typeof err.response.body === 'object') {
                    childErr = err.response.body
                } else {
                    childErr = { ...JSON.parse(err.response.body.toString()) }
                }
            } catch (_error) { /* do nothing */ }
            if (!childErr || typeof childErr !== 'object') {
                childErr = {}
            }
            Object.assign(niceError, childErr)
            niceError.message = childErr.error || childErr.message || niceError.message
            niceError.code = childErr.code || niceError.code
            niceError.stack = childErr.stack || niceError.stack
        }
    }
    if (statusCode === 413) {
        niceError.message = 'Quota exceeded.'
        if (childErr && childErr.limit) {
            niceError.message += ` The current limit is ${childErr.limit} bytes.`
        }
        niceError.code = 'quota_exceeded'
    }
    niceError.stack = niceError.stack || err.stack
    niceError.code = niceError.code || err.code
    niceError._normalised = true // prevent double processing
    return niceError
}

class FFContextStorage {
    constructor (opts) {
        opts = opts || {}
        const projectID = opts?.projectID || (process.env.FF_FS_TEST_CONFIG ? process.env.FLOWFORGE_PROJECT_ID : null)
        const projectToken = opts?.token || (process.env.FF_FS_TEST_CONFIG ? process.env.FLOWFORGE_PROJECT_TOKEN : null)
        const fileStoreURL = opts?.url || 'http://127.0.0.1:3001'
        this.validSetup = projectID && projectToken && fileStoreURL && fileStoreURL.startsWith('http')
        if (!this.validSetup) {
            console.warn(CONFIG_ERROR_MSG)
        }
        this.knownCircularRefs = {}
        /** @type {import('got').Got} */
        this.client = got.extend({
            prefixUrl: `${fileStoreURL}/v1/context/${projectID}`,
            headers: {
                'user-agent': 'FlowForge Node-RED File Nodes for Storage Server',
                authorization: 'Bearer ' + projectToken
            },
            timeout: {
                request: opts.requestTimeout || 500
            },
            retry: {
                limit: 0
            }
        })
        this.initialised = !!this.client
    }

    open () {
        return Promise.resolve()
    }

    close () {
        return Promise.resolve()
    }

    /**
     * Get one or more values from the context store
     * @param {'context'|'flow'|'global'} scope - The scope of the context to get keys for
     * @param {string|Array<string>} key - The key to get the value for
     * @param {Function} callback - The callback to call when the value has been retrieved
     * @example
     *     // get a single value
     *     http://localhost:3001/v1/context/project-id/flow?key=hello
     * @example
     *     // get multiple values
     *     http://localhost:3001/v1/context/project-id/global?key=hello&key=nested.object.property
     */
    get (scope, key, callback) {
        const path = `${scope}`
        const keys = Array.isArray(key) ? key : [key]
        if (!this.validSetup) {
            return eeRejectOrCallback(!callback, callback)
        }
        if (typeof callback !== 'function') {
            // TODO: consider adding a cache to permit synchronous calls
            return Promise.reject(new Error('This context store must be called asynchronously'))
        }
        const opts = {
            search: new URLSearchParams(keys.map(k => ['key', k])),
            responseType: 'json'
        }
        this.client.get(path, opts).then(res => {
            callback(null, ...reviver(keys, res.body))
        }).catch(error => { // TODO: If resource not found, return null or error?
            callback(normaliseError(error))
        })
    }

    /**
     * Set one or more values in the context store
     * @param {'context'|'flow'|'global'} scope - The scope of the context to set
     * @param {string|Array<string>} key - The key(s) to set the value for
     * @param {string|Array<string>} value - The value(s) to set for the given scope + key(s)
     * @param {Function} callback - The callback to call when the value(s) have been set
     * @example
     *    // set a single value
     *    http://localhost:3001/v1/context/project-id/flow
     *    // body
     *    [{ "key": "hello", "value": "world" }]
     * @example
     *    // set multiple values
     *    http://localhost:3001/v1/context/project-id/flow
     *    // body
     *    [{ "key": "hello", "value": "world" }, { "key": "nested.object.property", "value": "value" }]
     */
    set (scope, key, value, callback) {
        if (!this.validSetup) {
            return eeRejectOrCallback(!callback, callback)
        }
        if (typeof callback !== 'function') {
            // TODO: consider adding a cache to permit synchronous calls
            throw new Error('This context store must be called asynchronously')
        }
        const path = `${scope}`
        const data = Array.isArray(key) ? key.map((k, i) => ({ key: k, value: value[i] })) : [{ key, value }]
        const stringifiedContext = stringify(data)
        const opts = {
            responseType: 'json',
            body: stringifiedContext.json,
            headers: { 'Content-Type': 'application/json;charset=utf-8' }
        }
        this.client.post(path, opts).then(res => {
            callback(null)
        }).catch(error => {
            callback(normaliseError(error))
        })
    }

    /**
     * Get a list of keys for a given scope
     * @param {'context'|'flow'|'global'} scope - The scope of the context to get keys for
     * @param {Function} callback - The callback to call when the keys have been retrieved
     * @example
     *     http://localhost:3001/v1/context/project-id/global/keys
     */
    keys (scope, callback) {
        if (!this.validSetup) {
            if (callback) {
                callback(null, []) // quietly return empty key list
            } else {
                return Promise.resolve([]) // quietly return empty key list
            }
        }
        if (typeof callback !== 'function') {
            // TODO: consider adding a cache to permit synchronous calls
            throw new Error('This context store must be called asynchronously')
        }
        const path = `${scope}/keys`
        const opts = {
            responseType: 'json',
            headers: {
                'Content-Type': 'application/json;charset=utf-8',
                Accept: 'application/json'
            }
        }
        this.client.get(path, opts).then(res => {
            callback(null, res.body || [])
        }).catch(error => {
            callback(normaliseError(error))
        })
    }

    /**
     * Delete the context of the given node/flow/global
     * @param {String} scope - the scope to delete
     */
    delete (scope) {
        if (!this.validSetup) {
            return Promise.resolve() // quietly ignore
        }
        return this.client.delete(scope).then(() => {
            // done
        }).catch(error => {
            error.code ||= 'unexpected_error'
            // TODO: log error?
        })
    }

    /**
     * Delete any contexts that are no longer in use
     * @param {Array<string>} _activeNodes - a list of nodes still active
     */
    clean (_activeNodes) {
        if (!this.validSetup) {
            return Promise.resolve() // quietly ignore
        }
        const activeNodes = _activeNodes || []
        const opts = { json: activeNodes }
        return this.client.post('clean', opts).then(() => {
            // done
        }).catch(error => {
            error.code ||= 'unexpected_error'
            // TODO: log error?
        })
    }

    _export () {
        // TODO: needed? I think not looking through @node-red/runtime/lib/nodes/context/index.js
        return []
    }
}

module.exports = function (config) {
    return new FFContextStorage(config)
}


lib/FFContextStorage.js
const MemoryCache = require('./memoryCache')
const got = require('got').default
const safeJSONStringify = require('json-stringify-safe')
const CONFIG_ERROR_MSG = 'Persistent context plugin cannot be used outside of FlowForge EE environment'

/**
 * @typedef {Object} FFContextStorageConfig - The configuration object for the FlowForge Context Storage
 * @property {string} projectID - The FlowForge project ID
 * @property {string} token - The FlowForge project token
 * @property {string} [url='http://127.0.0.1:3001'] - The FlowForge File Store URL
 * @property {number} [requestTimeout=3000] - The number of milliseconds to wait before timing out a request
 * @property {number} [pageSize=20] - The number of context items/rows to fetch per page
 * @property {number} [flushInterval=30] - The number of seconds to wait before flushing pending writes
 * @property {boolean} [cache=true] - Whether to cache context items in memory (required for synchronous get/set)
 */

class FFContextStorage {
    constructor (/** @type {FFContextStorageConfig} */ config) {
        // Ensure sane config
        config = config || {}
        config.pageSize = Math.max(1, config.pageSize || 20)
        config.requestTimeout = Math.max(500, config.requestTimeout || 3000)
        config.flushInterval = Math.max(0, config.flushInterval || 30) * 1000
        config.cache = Object.hasOwn(config, 'cache') ? config.cache : true
        config.projectID = config?.projectID || (process.env.FF_FS_TEST_CONFIG ? process.env.FLOWFORGE_PROJECT_ID : null)
        config.token = config?.token || (process.env.FF_FS_TEST_CONFIG ? process.env.FLOWFORGE_PROJECT_TOKEN : null)
        config.url = config?.url || 'http://127.0.0.1:3001'

        // setup class vars
        this.validSetup = false
        /** @type {FFContextStorageConfig} */
        this.config = config
        if (config.cache) {
            this.cache = MemoryCache({})
        }
        this.pendingWrites = {}
        this.knownCircularRefs = {}
        this.writePromise = Promise.resolve()
        this._flushPendingWrites = function () { return Promise.resolve() } // place holder for later
        this.validSetup = config.projectID && config.token && config.url?.startsWith('http')
        if (!this.validSetup) {
            console.warn(CONFIG_ERROR_MSG)
        }

        // create HTTP client instance for this project
        /** @type {import('got').Got} */
        this.client = got.extend({
            prefixUrl: `${config.url}/v1/context/${config.projectID}`,
            headers: {
                'user-agent': 'FlowForge Node-RED File Nodes for Storage Server',
                authorization: 'Bearer ' + config.token
            },
            timeout: {
                request: config.requestTimeout
            },
            retry: {
                limit: 0
            }
        })
    }

    async open () {
        const self = this
        if (!self.validSetup) {
            return Promise.reject(new Error(CONFIG_ERROR_MSG))
        }
        if (!self.cache) {
            self._flushPendingWrites = function () { return Promise.resolve() }
            return Promise.resolve()
        }
        const opts = {
            responseType: 'json'
        }
        self.nextActiveCursor = null
        const limit = self.config.pageSize
        let cursor = null
        let result = await getNext(cursor, limit)
        while (result) {
            updateCache(result)
            cursor = result.meta?.next_cursor
            if (cursor) {
                result = await getNext(cursor, limit)
            } else {
                result = null
            }
        }

        async function getNext (cursor, limit) {
            const path = paginateUrl('cache', cursor, limit)
            const response = await self.client.get(path, opts)
            return response?.body
        }

        function updateCache (result) {
            const rows = result?.data || []
            for (let index = 0; index < rows.length; index++) {
                const row = rows[index]
                if (typeof row !== 'object') { continue }
                const scope = row.scope
                if (!scope) { continue }
                const values = row?.values
                if (typeof values !== 'object') { continue }
                Object.keys(values).forEach(function (key) {
                    self.cache.set(scope, key, values[key])
                })
            }
        }
        // update _flushPendingWrites to a real function
        self._flushPendingWrites = function () {
            const scopes = Object.keys(self.pendingWrites)
            self.pendingWrites = {}
            const promises = []
            const newContext = self.cache._export()
            scopes.forEach(function (scope) {
                const context = newContext[scope] || {}
                const stringifiedContext = stringify(context)
                if (stringifiedContext.circular && !self.knownCircularRefs[scope]) {
                    console.warn('context.flowforge.error-circular', scope)
                    self.knownCircularRefs[scope] = true
                } else {
                    delete self.knownCircularRefs[scope]
                }
                promises.push(self._writeCache(scope, stringifiedContext.json))
            })
            delete self._pendingWriteTimeout
            return Promise.all(promises)
        }
    }

    close () {
        const self = this
        if (this.cache && this._pendingWriteTimeout) {
            clearTimeout(this._pendingWriteTimeout)
            delete this._pendingWriteTimeout
            this.config.flushInterval = 0
            self.writePromise = self.writePromise.then(function () {
                return self._flushPendingWrites().catch(function (err) {
                    console.error('Error flushing pending context writes:' + err.toString())
                })
            })
        }
        return this.writePromise
    }

    /**
     * Get one or more values from the context store
     * @param {'context'|'flow'|'global'} scope - The scope of the context to get keys for
     * @param {string|Array<string>} key - The key to get the value for
     * @param {Function} callback - The callback to call when the value has been retrieved
     * @example
     *     // get a single value
     *     http://localhost:3001/v1/context/project-id/flow?key=hello
     * @example
     *     // get multiple values
     *     http://localhost:3001/v1/context/project-id/global?key=hello&key=nested.object.property
     */
    get (scope, key, callback) {
        if (!this.validSetup) {
            return eeRejectOrCallback(!callback, callback)
        }
        if (this.cache) {
            return this.cache.get(scope, key, callback)
        } else if (typeof callback !== 'function') {
            throw new Error('This context store must be called asynchronously')
        } else {
            const path = `${scope}`
            const keys = Array.isArray(key) ? key : [key]
            const opts = {
                search: new URLSearchParams(keys.map(k => ['key', k])),
                responseType: 'json'
            }
            this.client.get(path, opts).then(res => {
                callback(null, ...reviver(keys, res.body))
            }).catch(error => {
                callback(normaliseError(error))
            })
        }
    }

    /**
     * Set one or more values in the context store
     * @param {'context'|'flow'|'global'} scope - The scope of the context to set
     * @param {string|Array<string>} key - The key(s) to set the value for
     * @param {string|Array<string>} value - The value(s) to set for the given scope + key(s)
     * @param {Function} callback - The callback to call when the value(s) have been set
     * @example
     *    // set a single value
     *    http://localhost:3001/v1/context/project-id/flow
     *    // body
     *    [{ "key": "hello", "value": "world" }]
     * @example
     *    // set multiple values
     *    http://localhost:3001/v1/context/project-id/flow
     *    // body
     *    [{ "key": "hello", "value": "world" }, { "key": "nested.object.property", "value": "value" }]
     */
    set (scope, key, value, callback) {
        const self = this
        if (!self.validSetup) {
            return eeRejectOrCallback(!callback, callback)
        }
        if (self.cache) {
            self.cache.set(scope, key, value, callback)
            self.pendingWrites[scope] = true
            if (self._pendingWriteTimeout) {
                // there's a pending write which will handle this
            } else {
                self._pendingWriteTimeout = setTimeout(function () {
                    self.writePromise = self.writePromise.then(function () {
                        return self._flushPendingWrites().catch(function (err) {
                            console.error('Error flushing pending context writes:' + err.toString())
                        })
                    })
                }, self.flushInterval)
            }
        } else if (typeof callback !== 'function') {
            throw new Error('This context store must be called asynchronously')
        } else {
            const path = `${scope}`
            const data = Array.isArray(key) ? key.map((k, i) => ({ key: k, value: value[i] })) : [{ key, value }]
            const stringifiedContext = stringify(data)
            const opts = {
                responseType: 'json',
                body: stringifiedContext.json,
                headers: { 'Content-Type': 'application/json;charset=utf-8' }
            }
            self.client.post(path, opts).then(res => {
                callback(null)
            }).catch(error => {
                callback(normaliseError(error))
            })
        }
    }

    /**
     * Get a list of keys for a given scope
     * @param {'context'|'flow'|'global'} scope - The scope of the context to get keys for
     * @param {Function} callback - The callback to call when the keys have been retrieved
     * @example
     *     http://localhost:3001/v1/context/project-id/global/keys
     */
    keys (scope, callback) {
        if (!this.validSetup) {
            if (callback) {
                callback(null, []) // quietly return empty key list
            } else {
                return Promise.resolve([]) // quietly return empty key list
            }
        }
        if (this.cache) {
            return this.cache.keys(scope, callback)
        }
        if (typeof callback !== 'function') {
            throw new Error('This context store must be called asynchronously')
        }

        const path = `${scope}/keys`
        const opts = {
            responseType: 'json',
            headers: {
                'Content-Type': 'application/json;charset=utf-8',
                Accept: 'application/json'
            }
        }
        this.client.get(path, opts).then(res => {
            callback(null, res.body || [])
        }).catch(error => {
            callback(normaliseError(error))
        })
    }

    /**
     * Delete the context of the given node/flow/global
     * @param {String} scope - the scope to delete
     */
    delete (scope) {
        const self = this
        if (!self.validSetup) {
            return Promise.resolve() // quietly ignore
        }
        let cachePromise
        if (self.cache) {
            cachePromise = self.cache.delete(scope)
        } else {
            cachePromise = Promise.resolve()
        }
        delete self.pendingWrites[scope]
        return cachePromise.then(function () {
            return self.client.delete(scope)
        })
    }

    /**
     * Delete any contexts that are no longer in use
     * @param {Array<string>} _activeNodes - a list of nodes still active
     */
    clean (_activeNodes) {
        const self = this
        if (!self.validSetup) {
            return Promise.resolve() // quietly ignore
        }
        const activeNodes = _activeNodes || []
        const opts = { json: activeNodes }
        let cachePromise
        if (self.cache) {
            cachePromise = self.cache.clean(activeNodes)
        } else {
            cachePromise = Promise.resolve()
        }
        self.knownCircularRefs = {}
        return cachePromise.then(function () {
            return self.client.post('clean', opts)
        }).then(() => {
            // done
        }).catch(error => {
            error.code ||= 'unexpected_error'
            // TODO: log error?
        })
    }

    _export () {
        // TODO: needed? I think not looking through @node-red/runtime/lib/nodes/context/index.js
        return []
    }

    async _writeCache (scope, jsonData) {
        const self = this
        const path = `cache/${scope}`
        const opts = {
            responseType: 'json',
            body: jsonData,
            headers: { 'Content-Type': 'application/json;charset=utf-8' }
        }
        console.debug('writeCache scope ' + scope)
        const result = await self.client.post(path, opts)
        console.log('writeCache result', result)
    }
}

// #region helpers
function stringify (value) {
    let hasCircular
    const result = safeJSONStringify(value, null, null, function (k, v) { hasCircular = true })
    return { json: result, circular: hasCircular }
}

const reviver = (keys, data) => {
    const result = keys.map(key => {
        const el = data.find(e => e.key === key)
        return el?.value
    })
    return result
}

function eeRejectOrCallback (reject, callback) {
    if (callback) {
        callback(new Error(CONFIG_ERROR_MSG))
    } else if (reject) {
        return Promise.reject(new Error(CONFIG_ERROR_MSG))
    }
}

function normaliseError (err) {
    let niceError = new Error('Unexpected error.')
    let statusCode = null
    let childErr = {}
    niceError.code ||= 'unexpected_error'
    if (typeof err === 'string') {
        niceError = new Error(err)
    } else if (err?._normalised) {
        return err // already normalised
    }
    if (err?.response) {
        statusCode = err.response.statusCode
        if (err.response.body) {
            try {
                if (err.response.body && typeof err.response.body === 'object') {
                    childErr = err.response.body
                } else {
                    childErr = { ...JSON.parse(err.response.body.toString()) }
                }
            } catch (_error) { /* do nothing */ }
            if (!childErr || typeof childErr !== 'object') {
                childErr = {}
            }
            Object.assign(niceError, childErr)
            niceError.message = childErr.error || childErr.message || niceError.message
            niceError.code = childErr.code || niceError.code
            niceError.stack = childErr.stack || niceError.stack
        }
    }
    if (statusCode === 413) {
        niceError.message = 'Quota exceeded.'
        if (childErr && childErr.limit) {
            niceError.message += ` The current limit is ${childErr.limit} bytes.`
        }
        niceError.code = 'quota_exceeded'
    }
    niceError.stack = niceError.stack || err.stack
    niceError.code = niceError.code || err.code
    niceError._normalised = true // prevent double processing
    return niceError
}

function paginateUrl (url, cursor, limit, query) {
    const queryString = new URLSearchParams()
    if (cursor) {
        queryString.append('cursor', cursor)
    }
    if (limit) {
        queryString.append('limit', limit)
    }
    if (query) {
        queryString.append('query', query)
    }
    const qs = queryString.toString()
    if (qs === '') {
        return url
    }
    return `${url}?${qs}`
}
// #endregion helpers

exports.FFContextStorage = FFContextStorage
